app/grandchallenge/emails/emails.py

```diff
@@ -1,34 +1,111 @@
-    ):
-        # Do not send unsubscribable emails to users who have opted out
-        return
+def filter_recipients(recipients, subscription_type=None):
+    filtered_recipients = []
+    for recipient in recipients:
+        if not recipient.is_active:
+            # Do not send emails to inactive users
+            continue
 
-    send_mail(
+        if (
+            subscription_type == EmailSubscriptionTypes.NOTIFICATION
+            and not recipient.user_profile.receive_notification_emails
+        ) or (
+            subscription_type == EmailSubscriptionTypes.NEWSLETTER
+            and not recipient.user_profile.receive_newsletter
+        ):
+            # Do not send emails to users who have opted out
+            continue
+
+        filtered_recipients.append(recipient)
+
+    return filtered_recipients
+
+
+def send_standard_email_batch(
+    *, subject, message, recipients, subscription_type=None
+):
+    connection = get_connection()
+    messages = []
+    recipients = filter_recipients(
+        recipients=recipients, subscription_type=subscription_type
+    )
+    for recipient in recipients:
+        messages.append(
+            create_email_object(
+                recipient=recipient,
+                subject=subject,
+                message=message,
+                subscription_type=subscription_type,
+                connection=connection,
+            )
+        )
+    return connection.send_messages(messages)
+
+
+def get_unsubscribe_link(recipient, subscription_type):
+    if subscription_type == EmailSubscriptionTypes.NEWSLETTER:
+        return reverse(
+            "newsletter-unsubscribe",
+            kwargs={"token": recipient.user_profile.unsubscribe_token},
+        )
+    elif subscription_type == EmailSubscriptionTypes.NOTIFICATION:
+        return reverse(
+            "notification-unsubscribe",
+            kwargs={"token": recipient.user_profile.unsubscribe_token},
+        )
+    else:
+        return None
+
+
+def get_headers(unsubscribe_link):
+    if unsubscribe_link:
+        return {
+            "List-Unsubscribe": unsubscribe_link,
+            "List-Unsubscribe-Post": "List-Unsubscribe=One-Click",
+        }
+    else:
+        return None
+
+
+def create_email_object(
+    recipient, subject, message, connection, subscription_type=None
+):
+    site = Site.objects.get_current()
+    unsubscribe_link = get_unsubscribe_link(recipient, subscription_type)
+    headers = get_headers(unsubscribe_link)
+
+    html_content = render_to_string(
+        "vendored/mailgun_transactional_emails/action.html",
+        {
+            "title": subject,
+            "username": recipient.username,
+            "content": message,
+            "unsubscribe_link": unsubscribe_link,
+            "subscription_type": subscription_type,
+            "site": site,
+        },
+    )
+    html_content_without_linebreaks = html_content.replace("\n", "")
```

Comment:
How do the plain text emails look? Doesn't this approach strip out all paragraph breaks and links?

Did you consider the approach of writing all of our emails in Markdown, which is human readable, sending the rendered markdown as the html and the plain markdown as the plaintext?

Reply:
I changed all email messages to html and removed line breaks, so they look fine, but I agree that using markdown would be nicer. Thanks for the suggestion!

Reply:
Hm, looks like replacing of linebreaks is still necessary because the `md2email_html` function doesn't handle line breaks:

`First line\r\nSecond line` --> `<p>First line<br>\nSecond line</p>`
`First line\rSecond line` --> `<p>First line\nSecond line</p>`
`First line\nSecond line` --> `<p>First line\nSecond line</p>`

and the \n in there needs to go...

Reply:
We can circumvent this issue in our service emails by using the `<br/>` tag in the message instead of `\n`, but this is an issue for the newsletter emails since they will be written in the markdown editor and will result in the first example. So I suggest we keep `html_content_without_linebreaks = html_content.replace("\n", "")`.

Reply:
and using the markdown message as the plain text for the email would be incomplete (without username and the unsubscribe footer). We can do the conversion from md to html in the standard method, but I will keep all other steps in there.

Reply:
Are you sure your test is correct here?

If you type `\n` in a markdown file, then that is the literal `\` character followed by the `n` character, so will be rendered. If you put `\n` in a string that passes through the function, it is the newline character, so will not be rendered. It is no problem to send newline character in plain text emails. In fact, they should be there. In markdown, a paragraph break is represented by two newlines (`\n\n`), in which case a new paragraph break will be inserted.

We should not have carriage returns (`\r`) anywhere in our codebase as we use unix line endings.

Reply:
I tried this out by creating an email in the UI (like the 10% team would do for the newsletter):
![image](https://github.com/comic/grand-challenge.org/assets/30069334/f6fff58c-1d49-4635-8e7d-24f889522848)

then retrieving the email in the shell:
```
>>> email.body
'First line\r\n\r\nSecond line'`
>>> md2email_html(email.body)
'<p>First line</p>\n<p>Second line</p>'
```

The `\n` shouldn't be there, or at least we don't want it there because it will get inserted as is in the html email template and would look like this: 

![image](https://github.com/comic/grand-challenge.org/assets/30069334/c976255b-ac89-4c9d-a98a-57e24e18a696)

Reply:
For example, for the newsletter, you have `email.body`, which is a markdown string. You would then send this to `send_standard_email_batch` as `message=email.body` (maybe a better name for `message` is `message_markdown`).

Then, in `send_standard_email_batch` you generate `message_html = md2email_html(message_markdown)` and send out `message_html` and `message_markdown`.

For the other emails you create a template, e.g. app/grandchallenge/challenges/templates/challenges/partials/challenge_request_confirmation_email.md

```markdown
Your challenge request has been sent to the reviewers. You will receive an email informing you of our decision within the next 4 weeks.

The reviewers might contact you for additional information during that time. You can track the status of your request [here]({{ link }}).

We have calculated the following budget estimate. This estimate is based on the information you provided in the form and reflects a rough estimation of the costs we expect to incur:

{% for key, value in budget.items %}
    - {{ key }}: {{ value }} €
{% endfor %}
```

You then do:

```python
message_markdown = render_to_string(
         "challenges/partials/challenge_request_confirmation_email.md",
         context,
     )
```

For the other emails probably what we had before was almost valid markdown:

```python
 message_markdown = format_html(
         (
             "You have {unread_notification_count} new notification{suffix}.\n"
             "Read and manage your notifications [here]({url})."
         ),
         unread_notification_count=n_notifications,
         suffix=pluralize(n_notifications),
         url=reverse("notifications:list"),
     )
```

The `\n` in there is fine as plain text.

Reply:
Yes, the last two examples work (except for the fact that `message_markdown` does not contain the "Dear username" part and also not the unsubscribe footer, but in terms of linebreaks, all good.

But there is still the issue with the newsletter email, see my comment above.

Reply:
Why isn't this an issue with the current way of sending out the newsletter?

Reply:
> (except for the fact that message_markdown does not contain the "Dear username" part and also not the unsubscribe footer, but in terms of linebreaks, all good.

Ah of course, you would need a standard template to put it in to for that.

Reply:
> Why isn't this an issue with the current way of sending out the newsletter?

It is an issue with the current way of sending out the newsletter, which is why we do `html_content_without_linebreaks = html_content.replace("\n", "")` to get rid of the unnecessary line breaks in the html message.

Reply:
> Ah of course, you would need a standard template to put it in to for that.

Yes, so that's why I now take the rendered html message and turn that back into text with `strip_tags`. Otherwise, we would need two templates, one html and one plain text...

Reply:
The plain text is only added as a fall back if someone's email client doesn;t support html emails, so I think it would be fine to only include the core message there. The unsubscribe link is in the header of the email, so we would still fullfill the new requirements.

Reply:
Yes, but we want to avoid using `strip_tags` as that gets rid of all the links, right?

Reply:
> The plain text is only added as a fall back if someone's email client doesn;t support html emails, so I think it would be fine to only include the core message there. The unsubscribe link is in the header of the email, so we would still fullfill the new requirements.

We would not, the requirements are to have a prominent link at the footer of the email too.

Reply:
I still do not understand why `{{ object.body|md2email_html }}` works fine in `email_detail.html`, but we need extra processing when doing `md2email_html(object.body)` followed by `render_to_string(...)`. Why not do `{{ content|md2email_html }}` in `action.html`? Does it have the same problem?

Reply:
Doing `{{ content|md2email_html }}` in `action.html` actually seems to work!

Reply:
> We would not, the requirements are to have a prominent link at the footer of the email too.

Where do you read that? The RFCs only mention the headers. Of course, it is much user friendlier to add a link in the footer, and I can create a second template to use for the plain text content, but strictly-speaking, just adding the headers is enough to fullfill the requirements as far as I know. It's only if you don't have the headers, that Google and Yahoo will keep accepting the mail until June 2024 as long as you do have a prominent link in the footer.

Reply:
https://support.google.com/a/answer/81126?sjid=10053208798928542167-EU&visit_id=638442052921949914-1483516976&rd=1#zippy=%2Crequirements-for-sending-or-more-messages-per-day

Marketing messages and subscribed messages must support one-click unsubscribe, and include a clearly visible unsubscribe link in the message body.

Reply:
I think having a secondary standard `.txt` template that adds the greeting, content block, sign-off and unsubscribe link is fine. We should be able to keep the two synced and you will end up with better looking email than by stripping the tags.